### ISSUE TITLE
QUIC Datagram API

### ISSUE THREAD

**Issue by wegylexy** [dotnet/runtime#53533]
## Background and Motivation

QUIC is finally a proposed standard in RFC, with HTTP/3 and WebTransport on the way.
To prepare for WebTransport and other use cases, such as unreliable message delivery in SignalR, .NET should implement QUIC datagram API, as msquic already supports it, to enable higher-level APIs such as WebTransport.
Until WebTransport is standardized, it may be used today to stream real-time game state and ultra low-latency voice data where dropped packets should not be retransmitted. Once this is done, SignalR may support new fetures.

## Proposed API

```diff
namespace System.Net.Quic
{
    public class QuicClientConnectionOptions
    {
+        public bool DatagramReceiveEnabled { get { } set { } }
    }
+    public delegate void QuicDatagramReceivedEventHandler(object sender, ReadOnlySpan<byte> buffer);
    public class QuicConnection
    {
+        public bool DatagramReceivedEnabled { get { } set { } }
+        public bool DatagramSendEnabled { get { } set { } }
+        public ushort DatagramMaxSendLength { get { } }
+        public event QuicDatagramReceivedEventHandler? DatagramReceived { add { } remove { } }
+        public ValueTask<bool> SendDatagramAsync(ReadOnlyMemory<byte> buffer, bool priority = false) { }
    }
    public class QuicListenerOptions
    {
+        public bool DatagramReceiveEnabled { get { } set { } }
    }
}
```
See https://github.com/wegylexy/runtime/pull/1 for implementation with msquic (tested with v1.3.0).

## Usage Examples

``` C#
// receive
connection.DatagramReceived += (sender, buffer) =>
{
    // Parse the readonly span synchronously, without copying all the bytes, into an async task
    MyAsyncChannel.Writer.TryWrite(MyZeroCopyHandler.HandleAsync(buffer));
}
// send
var array = ArrayPool<byte>.Shared.Rent(Unsafe.SizeOf<MyTinyStruct>());
try
{
    MemoryMarshal.Cast<byte, MyTinyStruct>(array).SetCurrentGameState();
    await connection.SendDatagramAsync(array);
}
finally
{
    ArrayPool<byte>.Shared.Return(array);
}
```

## Alternative Designs

Receiving datagram buffers with a channel (similar to the stream API) was considered, but the msquic datagram buffer is merely a pointer to the underlying UDP buffer such that the buffer content is only available during the event callback. Async handler implies unnecessary cloning of possibly a thousand bytes and increase GC pressure for every single datagram received.

Sending with a readonly span was considered for `stackalloc`ed buffer, but msquic needs to hold on to the memory beyond the scope of the method.

Sending returns a bool to indicate whether the datagram is acknowledged without spurious loss, or throws when discarded/cancelled. Additional API may be added to handle suspected loss.

## Risks

`SendDatagramAsync()` could return something better than a simple `bool` when the task completes.

**Comment by wegylexy** [comment]
@karelz seems the bot has trouble adding `area-System.Net.Quic`

**Comment by ThadHouse** [comment]
One thing to note about the msquic side of this API is that DatagramMaxSendLength can dynamically change over the course of a connection. It currently can't go down (Although in the future that might change) but it can go up after the connection has been started. Just something to note.

**Comment by wegylexy** [comment]
@ThadHouse If notification is needed, could implement `INotifyPropertyChanged`. Otherwise, the getter could be invoked right before filling in the buffer to determine the current max.
See also [line 327](https://github.com/wegylexy/runtime/pull/1/files#diff-6b1b09f9599d7fce2389d0218dd65b7679d5290d6b87000ada08ee10e67cf072R327) where the msquic datagram-state-changed event handler updates exactly this property.

**Comment by ThadHouse** [comment]
I know theres a general rule to not have properties update without user interaction, which is the main reason I brought it up. Not a bug in the implementation, just a note.

**Comment by wegylexy** [comment]
Besides, `Connected` (in the current API) and `DatagramSendEnabled` (in the proposed API) may also update without user interaction, due to peer or transport.

**Comment by karelz** [comment]
Unless the RFC lands by March or earlier, we won't be able to react in time based on this year experience.

@wegylexy would you be interested in doing a prototype of the API as mentioned above? (Or we could run it by our API review when it is reasonably close to final shape) ... we might not be ready for that right now, but in couple of months it should be feasible from our side.

**Comment by wegylexy** [comment]
@karelz [wegylexy/runtime/.../feature/msquic-datagram](https://github.com/wegylexy/runtime/tree/feature/msquic-datagram) is a prototype, forked from a .NET 6 preview when `System.Net.Quic` was still `public`.

**Comment by wegylexy** [comment]
HTTP/3 WebTransport will need HTTP Datagram which requires QUIC Datagram.

**Comment by wegylexy** [comment]
Alternative design: https://github.com/StirlingLabs/MsQuic.Net/blob/main/StirlingLabs.MsQuic/QuicDatagram.cs

**Comment by wegylexy** [comment]
Updated to use `ReadOnlySequence<byte>` for zero-copy prefixing the datagram payload with a WebTransport session ID, which is another `ReadOnlyMemory<byte>` of 1-8 bytes.

A prototype is available at https://github.com/wegylexy/quic-with-datagram and a simple WebTransport server prototype on top of that is coming soon.

**Comment by wegylexy** [comment]
WebTransport server prototype: https://github.com/wegylexy/webtransport

**Comment by rzikm** [comment]
Since the QUIC Datagram extension (https://datatracker.ietf.org/doc/rfc9221/) has been officially released as RFC, is it worth trying to get it in for 7.0? @karelz

**Comment by nibanks** [comment]
Big +1 to exposing the datagram APIs in .NET. I'd love to see that happen in 7.0. Let me know if you need any help from the MsQuic side.

**Comment by karelz** [comment]
Given the load on the team and our capacity, I don't think we will be able to do anything meaningful in 7.0.

**Comment by rzikm** [comment]
Some interesting pieces from the RFC to consider during design:

>    Application protocols that use datagrams MUST define how they react
   to the absence of the max_datagram_frame_size transport parameter.
   If datagram support is integral to the application, the application
   protocol can fail the handshake if the max_datagram_frame_size
   transport parameter is not present.

=> We need to allow user code to decide whether the connection is aborted when Datagram support is absent

>   Note that while the max_datagram_frame_size transport parameter
   places a limit on the maximum size of DATAGRAM frames, that limit can
   be further reduced by the max_udp_payload_size transport parameter
   and the Maximum Transmission Unit (MTU) of the path between
   endpoints.  DATAGRAM frames cannot be fragmented; therefore,
   application protocols need to handle cases where the maximum datagram
   size is limited by other factors.

This is probably what ThadHouse meant when he said that the maximum datagram send size can change during the lifetime of the connection. 

>    QUIC implementations SHOULD present an API to applications to assign
   relative priorities to DATAGRAM frames with respect to each other and
   to QUIC streams.

So far we don't have any way to express prioritization of QuicStreams, so not sure how we would fit Datagrams in the scheme. @nibanks does MsQuic support stream/datagram prioritization?

> If a sender detects that a packet containing a specific DATAGRAM
   frame might have been lost, the implementation MAY notify the
   application that it believes the datagram was lost.
>
>    Similarly, if a packet containing a DATAGRAM frame is acknowledged,
   the implementation MAY notify the sender application that the
   datagram was successfully transmitted and received.  Due to
   reordering, this can include a DATAGRAM frame that was thought to be
   lost but, at a later point, was received and acknowledged. 

Putting here just to note that notifying user about datagram loss is not strictly required by the RFC.

**Comment by nibanks** [comment]
>  We need to allow user code to decide whether the connection is aborted when Datagram support is absent

We have the `QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED` notification that informs you if it's enabled or disabled.

> This is probably what ThadHouse meant when he said that the maximum datagram send size can change during the lifetime of the connection.

Correct. `QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED` also informs you of the `MaxSendLength`.

> So far we don't have any way to express prioritization of QuicStreams, so not sure how we would fit Datagrams in the scheme. @nibanks does MsQuic support stream/datagram prioritization?

For Streams, we have `QUIC_PARAM_STREAM_PRIORITY` which is just a `uint16_t` used to prioritize across streams. Currently, datagrams are always prioritized higher than streams.

> Putting here just to note that notifying user about datagram loss is not strictly required by the RFC.

We already have a notification for indicating loss, as a part of `QUIC_CONNECTION_EVENT_DATAGRAM_SEND_STATE_CHANGED`. It indicates:
```c
typedef enum QUIC_DATAGRAM_SEND_STATE {
    QUIC_DATAGRAM_SEND_UNKNOWN,                         // Not yet sent.
    QUIC_DATAGRAM_SEND_SENT,                            // Sent and awaiting acknowledegment
    QUIC_DATAGRAM_SEND_LOST_SUSPECT,                    // Suspected as lost, but still tracked
    QUIC_DATAGRAM_SEND_LOST_DISCARDED,                  // Lost and not longer being tracked
    QUIC_DATAGRAM_SEND_ACKNOWLEDGED,                    // Acknowledged
    QUIC_DATAGRAM_SEND_ACKNOWLEDGED_SPURIOUS,           // Acknowledged after being suspected lost
    QUIC_DATAGRAM_SEND_CANCELED,                        // Canceled before send
} QUIC_DATAGRAM_SEND_STATE;
```

**Comment by wegylexy** [comment]
> We need to allow user code to decide whether the connection is aborted when Datagram support is absent

https://github.com/wegylexy/runtime/blob/6abc64405576ba8b740ea1ca7ec9109b956b9455/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L949-L959

> This is probably what ThadHouse meant when he said that the maximum datagram send size can change during the lifetime of the connection.

https://github.com/wegylexy/runtime/blob/6abc64405576ba8b740ea1ca7ec9109b956b9455/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L512-L516

> does MsQuic support stream/datagram prioritization?

https://github.com/wegylexy/runtime/blob/6abc64405576ba8b740ea1ca7ec9109b956b9455/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L969-L986

> notifying user about datagram loss

https://github.com/wegylexy/runtime/blob/6abc64405576ba8b740ea1ca7ec9109b956b9455/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L997

**Comment by wegylexy** [comment]
Rebased my sample implementation onto .NET 7.0 https://github.com/wegylexy/runtime/commit/d4abf8e8fa0d09520d3d5651c41582c585673248

**Comment by alexrp** [comment]
> ```diff
> +        public System.Threading.Tasks.Task<QuicDatagramSendingResult> SendDatagramAsync(ReadOnlyMemory<byte> buffer, bool priority = false) { throw null; }
> +        public System.Threading.Tasks.Task<QuicDatagramSendingResult> SendDatagramAsync(System.Buffers.ReadOnlySequence<byte> buffers, bool priority = false) { throw null; }
> ```

The allocations implied by these APIs seems likely to be unacceptable to a lot of potential users. The `QuicDatagramSendingResult` class instance (and the tasks on it) should probably only be allocated if the user opts into doing so for a given call.

**Comment by wegylexy** [comment]
How about this?
```cs
public class DatagramSendOptions
{
    bool Priority { get { throw null; } set { } }
    Action? Sent { get { throw null; } set { } }
    Action? LostSuspect { get { throw null; } set { } }
    Action? Lost { get { throw null; } set { } }
    Action? Acknowledged { get { throw null; } set { } }
}

public Task SendDatagramAsync(ReadOnlyMemory<byte> buffer, DatagramSendOptions options = null);
public Task SendDatagramAsync(ReadOnlySequence<byte> buffer, DatagramSendOptions options = null);
```
When the task completes, it will be safe to release the buffer.

**Comment by wegylexy** [comment]
Edited proposed API just now to eliminate Task allocations for sending a datagram.

**Comment by safetonet-github** [comment]
Can anyone review the above for integration and give any feedback? I have significant interest in this and am willing to help. It looks like @wegylexy has done significant work and presently this appears to be on a "sometime in the future" milestone. Seems like an easy integration with high value for next release.

**Comment by rzikm** [comment]
I don't speak for the whole networking team, but I am interested in this feature as well. We really wanted to take this in 8.0, but we were a bit short-staffed and had to cut some features. We will probably have enough bandwidth to work on this in 9.0 timeframe.

**Comment by rzikm** [comment]
Related: https://github.com/microsoft/msquic/issues/3906

**Comment by Kation** [comment]
Still no due date?

**Comment by wegylexy** [comment]
I guess I'm switching over to Rust from .NET then.

**Comment by nibanks** [comment]
@wegylexy I'm curious what your use case for datagrams are. Since we added them at the msquic layer, I've come to the conclusion that Streams are still better in most cases. Many times, stuff you want to send unreliable is still bigger than MTU. So, if you use streams, but then abort them before they retransmit (we added a feature for this in msquic), then you effectively get datagram behavior with segmentation/reassembly. At least in MsQuic, streams are very light weight...

**Comment by wegylexy** [comment]
@nibanks For very high frequency aircraft position updates between Microsoft Flight Simulator and some other simulators, each datagram carries latitude, longitude, altitude, and some animation data at slightly lower frequency. Usually smaller than a hundred bytes, well below MTU. Push-to-talk voice communication is over reliable streams with high priority, and text messages low priority. The most complex part is that data are routed based on spatial proximity inside the virtual world instead of physical real world, so .NET garbage collection was nice that I didn't have to deal with memory management in C++ (in case of using MsQuic directly), but now Rust is growing fast. It would be easier for us to go that way rather than waiting for NativeAOT to integrate with native clients. Moreover, many users refuse to upgrade from Windows 10, but .NET QUIC checks OS version; too many hacks for us to get things installed.

**Comment by wegylexy** [comment]
> > now Rust is growing fast
> 
> Could you clarify this? Are you referring to the work on the MsQuic Rust crate, or some other Rust QUIC crate?

Oops, sorry. I'm referring to the crate [wtransport](https://github.com/wegylexy/wtransport/tree/patch-1) which handles all the H3 stuff. Moving up from low-level QUIC to WebTransport, so we can create web clients in JavaScript too.

**Comment by neuecc** [comment]
I'm interested in using this for games.
In games, we traditionally use UDP-based protocols to continuously send small data like coordinates.
We already have deployed a game communication framework that works with gRPC - HTTP/2 - TCP,
but we want to update the protocol using QUIC and add an option for no retransmission.

Can we build an API that correctly and efficiently handles operations like `abort them before they retransmit` in the C# API layer, regardless of the primitive MsQuic layer...?
Otherwise, I think Datagram would have simpler behavior.

In any case, .NET 10 is coming soon.
I'd like to know the plans.

**Comment by ManickaP** [comment]
> In any case, .NET 10 is coming soon.

We're already past feature complete, so no new features / APIs for .NET 10.

**Comment by wes-sleeman** [comment]
> There seems to be increasing interest, I'm gonna put this tentatively in 11.0. Note that as this is a new feature, it cannot be back ported.

@ManickaP Is this the sort of feature where a well-written implementation in a pull request would make it more likely to be adopted or is it best we make an API proposal and let it sit until feature grooming happens and a definitive goal-set for 11.0 is decided?

**Comment by wes-sleeman** [comment]
Per comments in the other issue, I'm moving my API proposal below. I'm also happy to take on the implementation. 

### Background and motivation

With the formal adoption of the unreliable datagram extension to the QUIC protocol under [RFC 9221](https://datatracker.ietf.org/doc/rfc9221/), many applications (and their associated protocols) have begun using them in situations requiring secure transfer of real-time or latency sensitive data. [The linked RFC](https://datatracker.ietf.org/doc/rfc9221/) provides a more detailed justification of the underlying feature.

This datagram extension is supported by the underlying MsQuic library already used to implement the rest of `System.Net.Quic`, but the datagram features are a notable miss. Adding them to the `QuicConnection` class provides a simple, intuitive, and frictionless means for developers to leverage their existing QUIC connections to take advantage of datagrams and aligns with the implementation in MsQuic itself.

### API Proposal

```diff
namespace System.Net.Quic;

+ [Flags]
+ public enum SendDatagramOptions
+ {
+     None = 0,
+     Allow0Rtt = 1,
+     DatagramPriority = 8,
+     CancelOnBlocked = 64
+ }

public class QuicConnection
{
+     public event EventHandler<DatagramReceivedEventArgs>? DatagramReceived;
+     public void SendDatagram(byte[] buffer, int offset, int count)
+     public void SendDatagram(byte[] buffer, int offset, int count, SendDatagramOptions options)
+     public void SendDatagram(ReadOnlyMemory<byte> buffer)
+     public void SendDatagram(ReadOnlyMemory<byte> buffer, SendDatagramOptions options)
}

+ public sealed partial class DatagramReceivedEventArgs : EventArgs
+ {
+     public required byte[] Buffer { get; init; }
+ }
```

All the above are either new types or straight additions to the existing `QuicConnection` class. No modification of existing behaviour is required _EXCEPT_ setting MsQuic param `QUIC_PARAM_CONN_DATAGRAM_RECEIVE_ENABLED` (probably via modifying the default `QUIC_SETTINGS` to enable `DatagramReceiveEnabled`) to indicate datagram support in new connections (i.e. in `ConnectAsync` and in the listener's `AcceptConnectionAsync`).

Note that all the send methods are synchronous. This feels odd to me, but aligns with the underlying implementation and makes sense in light of the proposed `DatagramPriority` flag (existing underlying MsQuic flag name `QUIC_SEND_FLAG_DGRAM_PRIORITY`) which ensures that the provided datagram is sent before others.

### API Usage

```csharp
using System.Net.Quic;
using System.Text;

// First, check if QUIC is supported.
if (!QuicConnection.IsSupported)
{
    Console.WriteLine("QUIC is not supported, check for presence of libmsquic and support of TLS 1.3.");
    return;
}

// This represents the minimal configuration necessary to open a connection.
QuicClientConnectionOptions clientConnectionOptions = new()
{
    // … Elided for brevity. See https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/quic/quic-overview …
};

// Initialize, configure and connect to the server.
var connection = await QuicConnection.ConnectAsync(clientConnectionOptions);
Console.WriteLine($"Connected {connection.LocalEndPoint} --> {connection.RemoteEndPoint}");

// For example, assume any received datagram is a UTF8-encoded text message.
// This also demonstrates the value of the event handler approach by showing how
// the connection can be recovered in a useful way, though here we could've closed-in
// the `connection` object as well.
connection.DatagramReceived += (conn, e) => {
    string message = Encoding.UTF8.GetString(e.Buffer);
    Console.WriteLine($"Received message from {conn.RemoteEndPoint}: {message}");
};

// Easy, simple helper function to send UTF8-encoded text messages.
// Could be replaced with any custom serialisation logic or other method
// of generating bytes.
void SendMessage(string message)
{
    // Serialise to a byte array and send that.
    // Could equivalently be done with a Memory<byte> instance.
    byte[] buffer = Encoding.UTF8.GetBytes(message);
    connection.SendDatagram(buffer, 0, buffer.Length);
}

// Send a few messages using the above function.
SendMessage("Hello, world!");
SendMessage("Answer life, the universe, and everything.");

// Example teardown.
// See https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/quic/quic-overview
await connection.CloseAsync(0x0C);
await connection.DisposeAsync();
```


### Alternative Designs

A more "full-fat" proposal was made before .NET 7.0 was released (see #53533), the discussion on which led to this more trimmed-down proposal.

Other options for end-developers include referencing MsQuic themselves directly, but this would require additional state management, etc, which `System.Net.Quic` is already doing under the hood.

Alternative "flavours" of this design include:
* Adding an internally maintained buffer of messages that can be pulled from (à la `System.Net.Sockets.TcpListener.AcceptSocketAsync`) at the cost of storing those messages; or
* Having a function to register a callback (though this is effectively re-inventing events); or
* Adding a boolean flag to `QuicConnectionOptions` which allows/requires clients to opt-in/opt-out (depending on default) to datagram support; or
* Changing the `SendDatagram` methods to `SendDatagramAsync` and returning a `ValueTask.CompletedTask` to make the API feel more natural.

One piece of additional functionality was considered but not included in the primary recommendation for the sake of simplicity:
* Adding a get-only boolean property to `QuicConnection` indicating if the connected remote supports datagrams (e.g. `DatagramSupported`) by querying MsQuic's `QUIC_PARAM_CONN_DATAGRAM_SEND_ENABLED` param.

### Risks

As written, the proposal necessitates marking datagrams as supported (passing `max_datagram_frame_size`) on all new connections which might cause some remotes to attempt to use said datagrams when communicating with .NET programs where they may not have before. The risk of this is fairly remote as RFC 9221 demands that any protocols which support datagrams provide an explanation of how they handle the lack of datagram support meaning any usage which could be impacted by this will likely be easily fixed by simply adding an appropriate datagram event handler.

Reliable transport is not typically a good fallback for an application which prefers unreliable transport, so this issue is unlikely to manifest in any meaningful way and requiring the flag to be set to enable datagrams will likely cause more person-hours of suffering than it saves.

The only other thing is less of a risk and more of a smell: The proposed functions are synchronous to match MsQuic and ensure the `DatagramPriority` flag has its desired effect. While this is correct, it _feels_ odd when the vast majority of other networking functions are asynchronous. I wouldn't be adverse to the last alternative design for the sake of developer familiarity, but I opted to keep the synchronous format as unnecessary asynchrony can limit the contexts in which the function can be used.

**Comment by ManickaP** [comment]
Keep this open, we're re-evaluating this for the next .NET. It's mandatory for WebTransport over H/3 AFAIK. And there's been quite a bit of activity on this issue over time. So this got a bit higher priority.

**Comment by Mrgaton** [comment]
Please 🥺 I want to make a virtual lan and using datagrams is a must, hope they implement it in next preview

**Comment by LAsBallenasSiempreSonPoderosas** [comment]
NOOOOO WHY NET 12.00 😭
